core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -181,15 +212,63 @@ public Schema getSchema() {
       return SCHEMA;
     }
 
+    public String extractPasswordFromUrl() {
+      int getLastIndexOfAtSign = path.lastIndexOf("@");
+      String authentication = path.substring(0, getLastIndexOfAtSign);
+      return authentication.substring(authentication.lastIndexOf(":") + 1);
+    }
+
+    public boolean authContainsSpecialCharacters() {
+      Pattern regularPasswordWithoutSpecialCharacters = Pattern.compile("[^A-Za-z0-9]");
+      Matcher regularPassword = regularPasswordWithoutSpecialCharacters.matcher(extractPasswordFromUrl());
+      return !regularPassword.matches();
+    }
+
     Map<String, String> getFileSystemProperties() {
-      if (fileSystemProperties == null) {
+      if (fileSystemProperties == null && !authContainsSpecialCharacters()) {
         return new HashMap<>();
       }
+      HashMap<String, String> fileSystemPropertiesMap = new HashMap<>();
+      if (fileSystemProperties != null) {
+        fileSystemPropertiesMap.putAll(GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE));
+      }
+
       try {
-        return GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE);
+        if (authContainsSpecialCharacters()) {
+          Path urlInfo;
+          String extractedPassword = extractPasswordFromUrl();
+          String encodedPassword = URLEncoder.encode(extractedPassword);
+          String validatePath = path.replace(extractedPassword, encodedPassword);
+          try {
+            urlInfo = new Path(validatePath);
+          } catch (Exception e) {
+            throw new IllegalArgumentException(String.format("Unable to parse url: %s %s", e.getMessage(), e));
+          }
+          int port = urlInfo.toUri().getPort();
```

Comment:
For Example: 

Before encoding: 
`ftp://kimi:42@67]gfuss@192.168.0.179:21/kimi-look-here.txt`
**This will throw an IllegalArgumentException because of invalid characters in password.** 

After encoding: 
`ftp://kimi:42%4067%5Dgfuss@192.168.0.179:21/kimi-look-here.txt`
**The Path (_org.apache.hadoop.fs.Path_) will be able to parse the url.**